### ISSUE TITLE
sod: premature shutdown for recurrence format 4

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#5945]
Cylc 8 may silently ignore some format 4 recurrences.

E.G:

```ini
[scheduler]
    cycle point format = CCYY
    allow implicit tasks = True

[scheduling]
    initial cycle point = 2000
    final cycle point = 2010
    [[graph]]
        # cycle forwards
        R3/2000/P1Y = f[-P1Y] => f

        # cycle backwards
        R3/P1Y/2010 = b[-P1Y] => b
```

**Cylc 8:** Spawns the `f` chain of tasks and shuts down on cycle 2003
**Cylc 7:** Spawns both the `f` and `b` chains.

**Comment by hjoliver** [comment]
Just tried with the "backward" recurrence alone:
```
INFO - Cylc version: 8.3.0.dev
INFO - Run mode: live
INFO - Initial point: 2000
INFO - Final point: 2010
INFO - Cold start from 2000
INFO - New flow: 1 (original flow from 2000) 2024-01-31 00:41:23
DEBUG - Runahead: base point 2008
DEBUG - Runahead limit: 2010
```

So this is runahead-related in the sense that at start-up, when there are no tasks in the pool, we compute the runahead limit using sequence points alone - something going wrong there it seems.

**Comment by MetRonnie** [comment]
Looks like it is a problem with the pre-initial dependency (if that's the right term). Adding in `R1//2007 = b` or `R1/2007 = b` causes the sequence to start

**Comment by hjoliver** [comment]
Yeah, I think I understand it ...

**Comment by hjoliver** [comment]
Actually, I'm kinda surprised this worked in Cylc 7. (Did it?)

The pre-initial dependency handling only ever (as I recall) applied to the initial cycle point of the workflow, not individual sequence start points. 

That rings a bell ...

**Comment by hjoliver** [comment]
Looks like I flagged this ages ago: 

https://github.com/cylc/cylc-flow/issues/1936

And see the final comment from 1 year ago:

https://github.com/cylc/cylc-flow/issues/1936#issuecomment-1031825906

**Comment by MetRonnie** [comment]
Ah yes, if you change the start of the format 3 recurrence (the "forwards" example) to a point beyond the ICP e.g. `R3/2002/P1Y = f[-P1Y] => f`, then you get the same immediate shutdown

**Comment by hjoliver** [comment]
Right, so as I understand it that's the expected result - hence my issue above suggesting it would nice to apply "pre-initial" logic to individual sequences not just the workflow initial cycle point.

The "workaround" is simply to not expect magical bootstrapping into an intercycle dependency, but handle it explicitly.

**Comment by hjoliver** [comment]
Explicit works fine:
```
    [[graph]]
        # count backwards
        R3/P1Y/2010 = b[-P1Y] => b
        # boostrap into the sequence
        R1/2007/P0Y = b
```

**Comment by oliver-sanders** [comment]
> Actually, I'm kinda surprised this worked in Cylc 7. (Did it?)

Yes (as above)

**Comment by hjoliver** [comment]
(Anyhow, I gotta bail, it's late here ... I'll check follow-up comments in the morning).

**Comment by hjoliver** [comment]
Cylc 7 stall with unsatisfied pre-spawned tasks, vs Cylc 8 shutdown with nothing to do - is expected under the circumstances - but I presume by "it worked at Cylc 7" you mean it actually ran, not that it immediately stalled.

**Comment by oliver-sanders** [comment]
> Well the lone "backward" one doesn't work:

That's because of the pre-initial dependency which I've written up as a separate issue, see https://github.com/cylc/cylc-flow/issues/5946

**Comment by MetRonnie** [comment]
So to uncross wires (as #5946 is the same issue as this one), is the _real_ issue here #4638?

**Comment by oliver-sanders** [comment]
I haven't had the time to investigate this yet to know.

**Comment by hjoliver** [comment]
OK, sorry if I didn't read the fine print:

> Cylc 7: Spawns both the f and b chains.

I guess I over-interpreted this to mean you were saying both chains actually **run** in Cylc 7.

We should certainly consider making this more obvious or flexible for users (hence the old issues #1936 and #4638) but technically the current Cylc 8 behaviour is correct and not a bug.

```
  R3/P1Y/2010 = b[-P1Y] => b  # with ICP = 2000 say
```
This literally says:
- spawn `2008/b` IF `2007/b` succeeds
- AND that `b` does not exist for points < `2008`

Plus:
- automatic bootstrapping into an inter-cycle dependency only applies to the ICP, not individual recurrences
- and you don't have to rely on automatic bootstrapping, even for the ICP

Therefore, under current well-defined rules of engagement the user has simply made a configuration error such that there is literally nothing to run in that part of the graph.

So on that basis this is a "could be better" rather than a bug, and once we've decided on the approach we should consolidate this issue with one of the older ones.

**Comment by oliver-sanders** [comment]
Follow up question:

Should we even allow a workflow to be started if it contains unreachable sections of graph?

Note, all functional Cylc 7 workflows will pass this test.

**Comment by hjoliver** [comment]
> Should we even allow a workflow to be started if it contains unreachable sections of graph?

Probably not, but (obviously) achieving that requires being able to detect that at validation.

> All functional Cylc 7 workflows will pass this test.

Which test?  You can certainly start a Cylc 7 workflow that contains unreachable graph: any inter-cycle dependence that isn't automatically bootstrapped by pre-initial-ignore:
```ini
[scheduling]
   cycling mode = integer
   initial cycle point = 1
   [[dependencies]]
      [[[R1]]]
              graph = "foo[-P1] => foo"  # OK
          [[[R1/2/P1]]]
              graph = "bar[-P1] => bar"  # Uh-oh, unreachable.
```

**Comment by oliver-sanders** [comment]
>> All **functional** Cylc 7 workflows will pass this test.
>
> Which test?

The test of not having an unreachable tasks.

By **functional** I was ruling out workflows with broken graphs that will stall when run.
